src/Components/Components/src/Services/UriHelperBase.cs

```diff
@@ -19,7 +19,7 @@ public abstract class UriHelperBase : IUriHelper
         {
             add
             {
-                EnsureInitialized();
+                AssertInitialized();
```

Comment:
I find the naming of these methods pretty confusing! There are now three initialization methods: `AssertInitialized`, `virtual EnsureInitialized`, `virtual InitializeState`. Of these, `AssertInitialized` doesn't just assert but tries to cause initialization also. Also it's unclear which of the two virtual methods a subclass should override in which scenarios.

How could this be simplified?

Reply:
I can't find a good way to do this, but also I've come to realize that it doesn't really matter. This is only visible to you when you override UriHelperBase to add your own implementation. Which is already a non starting point in some of our scenarios.

All our rendering scenarios require a specific UriHelperBase subtype (WASM requires WebAssemblyUriHelper, server-side requires RemoteUriHelper (it relies on the instance being RemoteUriHelper) and static rendering also uses its own implementation or falls back into RemoteUriHelper.

The only reason these classes are public is because they require a public static method that is [JSInvokable] in order to receive updates when the location changes and update the component accordingly.

I believe that we should at some point extract that logic from the UriHelpers and simply place it somewhere else and make the different IUriHelper implementations internal.

I can't think of a good scenario for extending any of these things as they are just implementations.

To put it in a different prism. UriHelperBase is using inheritance to share implementation, we could make this class internal and share the implementation with internalsvisibleto or simply share the same static class by linking the file in the two or three projects that need it and we would be better off.

If you agree with either:
1) It doesn't matter because its already very specific to each of the scenarios and not really replaceable, it's not in your face unless you downcast IUriHelper to UriHelperBase or try to extend the class, which both are not common.
2) We are just sharing an implementation through a base class and there are better ways to achieve this as described above.

Lets simply table this discussion by
1) Not doing anything because it doesn't matter.
2) Filing a bug and cleaning up this in the future by changing the way we share the core logic between our implementations.

Whatever you prefer I'm fine with both options.

Reply:
UriHelper should to be redesigned. The implementation has changed in ever PR that touches server side. Let's log an issue and if there are small wins now we should take them.

Reply:
Filed https://github.com/aspnet/AspNetCore/issues/8985